File: FinLib.NET/FinLib/Curve/Curve.cs
using System;
using FinLib.Extensions;
using FinLib.Price;

namespace FinLib.Curve;

public enum CurveType
{
    Absolute, Differential
}

public class Curve: IDisposable
{
    private readonly unsafe Curve_native* _curve;
    internal unsafe Curve_native* GetPtr() => _curve;

    public Curve(CurveType type)
    {
        unsafe
        {
            _curve = NativeMethods.curve_new(type.MapType());
        }
    }

    public nuint Count {
        get
        {
            unsafe
            {
                return NativeMethods.curve_size(_curve);
            }
        }
    }

    public void Add(double bid, double offer, DateTime date)
    {
        unsafe
        {
            NativeMethods.curve_add_rate_from(_curve, bid, offer, date.Year, Convert.ToUInt32(date.Month), Convert.ToUInt32(date.Day));
        }
    }

    public PricePair? RateAt(DateTime date)
    {
        unsafe
        {
            var item = NativeMethods.curve_get_rate(_curve, date.Year, Convert.ToUInt32(date.Month), Convert.ToUInt32(date.Day));
            if (item is null)
            {
                return null;
            }
            else
            {
                return new PricePair(item);
            }
        }
    }

    public PricePair? CumulativeRateAt(DateTime date)
    {
        unsafe
        {
            var item = NativeMethods.curve_get_cumulative_rate(_curve, date.Year, Convert.ToUInt32(date.Month), Convert.ToUInt32(date.Day));
            if (item is null)
            {
                return null;
            }
            else
            {
                return new PricePair(item);
            }
        }
    }

    public PricePair? AbsoluteRateAt(DateTime date)
    {
        unsafe
        {
            var item = NativeMethods.curve_get_absolute_rate(_curve, date.Year, Convert.ToUInt32(date.Month), Convert.ToUInt32(date.Day));
            if (item is null)
            {
                return null;
            }
            else
            {
                return new PricePair(item);
            }
        }
    }

    public PricePair CarryRate(DateTime from, DateTime to)
    {
        unsafe
        {
            var item = NativeMethods.curve_get_carry_rate(_curve,
                from.Year, Convert.ToUInt32(from.Month), Convert.ToUInt32(from.Day),
                to.Year, Convert.ToUInt32(to.Month), Convert.ToUInt32(to.Day)
                );
            return new PricePair(item);
        }
    }

    private void ReleaseUnmanagedResources()
    {
        unsafe
        {
            NativeMethods.curve_destroy(_curve);
        }
    }

    public void Dispose()
    {
        ReleaseUnmanagedResources();
        GC.SuppressFinalize(this);
    }

    ~Curve()
    {
        ReleaseUnmanagedResources();
    }
}

File: finlib-cpp/src/finlib/curve/curve.cpp
//
// Created by Andy Pack on 19/06/2025.
//

#include "curve.hpp"

namespace finlib {
    Curve::Curve(finlibrs::CurveType type) {
        handle = curve_new(type);
    }

    Curve::~Curve() {
        finlibrs::curve_destroy(handle);
    }

    size_t Curve::size() const {
        return finlibrs::curve_size(handle);
    }

    void Curve::add_rate_from(double bid, double offer, std::chrono::year_month_day &date) {
        finlibrs::curve_add_rate_from(handle, bid, offer, static_cast<int32_t>(date.year()),
                                      static_cast<uint32_t>(date.month()), static_cast<uint32_t>(date.day()));
    }

    std::unique_ptr<PricePair> Curve::get_rate(std::chrono::year_month_day &date) {
        auto price_handle = finlibrs::curve_get_rate(handle, static_cast<int32_t>(date.year()),
                                                     static_cast<uint32_t>(date.month()),
                                                     static_cast<uint32_t>(date.day()));
        if (price_handle) {
            return std::make_unique<PricePair>(price_handle);
        } else {
            return nullptr;
        }
    }

    std::unique_ptr<PricePair> Curve::get_absolute_rate(std::chrono::year_month_day &date) {
        auto price_handle = finlibrs::curve_get_absolute_rate(handle, static_cast<int32_t>(date.year()),
                                                              static_cast<uint32_t>(date.month()),
                                                              static_cast<uint32_t>(date.day()));
        if (price_handle) {
            return std::make_unique<PricePair>(price_handle);
        } else {
            return nullptr;
        }
    }

    std::unique_ptr<PricePair> Curve::get_cumulative_rate(std::chrono::year_month_day &date) {
        auto price_handle = finlibrs::curve_get_cumulative_rate(handle, static_cast<int32_t>(date.year()),
                                                                static_cast<uint32_t>(date.month()),
                                                                static_cast<uint32_t>(date.day()));
        if (price_handle) {
            return std::make_unique<PricePair>(price_handle);
        } else {
            return nullptr;
        }
    }

    std::unique_ptr<PricePair>
    Curve::get_carry_rate(std::chrono::year_month_day &from, std::chrono::year_month_day &to) {
        auto price_handle = finlibrs::curve_get_carry_rate(handle, static_cast<int32_t>(from.year()),
                                                           static_cast<uint32_t>(from.month()),
                                                           static_cast<uint32_t>(from.day()),
                                                           static_cast<int32_t>(to.year()),
                                                           static_cast<uint32_t>(to.month()),
                                                           static_cast<uint32_t>(to.day()));
        if (price_handle) {
            return std::make_unique<PricePair>(price_handle);
        } else {
            return nullptr;
        }
    }
}
